Docs/Configuration.tex

```diff
@@ -2573,6 +2573,17 @@ \subsection{Quirks Properties}\label{kernelpropsquirks}
   on all systems but those explicitly dedicated to scientific or media calculations.
   Only certain Xeon models typically benefit from the patch.
 
+\item
+  \texttt{CustomPciSerialDevice}\\
+  \textbf{Type}: \texttt{plist\ boolean}\\
+  \textbf{Failsafe}: \texttt{false}\\
+  \textbf{Requirement}: 10.7\\
```

Comment:
Since 10.4-10.6 looks way more complex, I do not want to bother with those until requested.

Reply:
@joevt Maybe you know more details about this patch than I do. Would you please try writing the docs?

Reply:
I don't know how to write OpenCore Docs.

Which parts do you need more explanation for?

**CustomPciSerialDevice**

This option causes the kernel (mostly pe_serial.c) to be patched to use the serial port specified by the Serial section. The patch changes all `in` and `out` accesses to 0x3f8-0x3ff (COM1 built-in serial port) to an address range described by the values in the Serial section.

There are two types of serial port devices each with a different calculation for the patched base address:
1) A PCI device specified by Serial→PciDeviceInfo. The base address for the address range is the value of the first I/O BAR (base address register) of the specified PCI device plus the value of Serial→RegisterBase (usually zero for a PCI device).
2) A non-PCI device (where Serial→PciDeviceInfo is a zero length path terminated by 0xff). The base address is the value of Serial→RegisterBase. For example, COM2 may exist at 0x2F8.

For the patch to be applied, the following must be true:
- Serial→Init must be true so that all the values in the Serial section (including PciDeviceInfo) are loaded into Pcd's.
- Serial→UseMmio must be false since the patch is only for devices controlled by port mapped I/O (PMIO).
- The calculated base address must not be the default 0x3f8.
The other values in the Serial section do not affect the kernel patch but will affect serial output from OpenCore when Debug→Target has bit 3 "Enable serial port logging" set.
- boot-args should have all the values required for the kernel to perform serial output. For example:
`debug=0x108 serial=3 serialbaud=115200`
- serialbaud in boot-args should match Serial→BaudRate if you're outputting to serial port in both OpenCore and the kernel.
- For kernel serial output to a non-default serial port device to work, you must use a method to stop a kext from matching the device. One method is to inject a codeless kext that matches the device using a higher IOProbeScore. A kext with code may be required if the normal matching methods cause the codeless kext to match multiple serial ports. For an example, search for `PCIeSerialDisable.kext/Contents/Info.plist` in https://github.com/acidanthera/bugtracker/issues/1954

Reply:
Thanks. I will try to write a more complete doc.

@vit9696 suggested we reduce module/section interdependency. Your points will be worth mentioning in the docs, but will not be how the patch works. Instead, we just apply it, though in most cases it does nothing (e.g. when it is an MMIO address which is greater than `UINT16_MAX`, it will not be applied; we have excluded 0x3F8 as well).

As for boot-args, can we actually set/append them automatically as we do for e.g. KASLR?

Reply:
> Thanks. I will try to write a more complete doc.

What did I miss?

> @vit9696 suggested we reduce module/section interdependency. Your points will be worth mentioning in the docs, but will not be how the patch works.

What are the differences between how the patch works and what I described?

> Instead, we just apply it, though in most cases it does nothing (e.g. when it is an MMIO address which is greater than `UINT16_MAX`, it will not be applied; we have excluded 0x3F8 as well).

I don't think the PCI spec requires PMIO addresses to be limited to 16 bits - but the patches we're making are limited to 16 bits because the instructions we are replacing only load 16 bits. That's ok because x86 architecture can't access PMIO addresses beyond 64K and the firmware and OS will ensure that the I/O BARs are always only 16 bits.

> EDIT - Currently we only exclude 0x3F8. Shall we exclude 0x3F8-0x3FF as well?

No. 0x3F8 is the base address of the addresses we are replacing. We replace all occurrences of any address between the base address and the base address plus 7. If the base address matches the base address used by the kernel (0x3F8) then we don't do anything. If the base address is different than the base address used by the kernel (for example 0x3FA) then we would have to replace all addresses in the range 0x3F9..0x401. But the base address will never be between 0x3F9 and 0x3FF because it must be a multiple of the size.

> As for boot-args, can we actually set/append them automatically as we do for e.g. KASLR?

My idea was to patch `PE_parse_boot_arg` to hide the fact that we're overriding the boot-args separately from what is in the boot-args string.

Regarding the PMIO patches, I think we need to use the `+` operator instead of the `|` operator because we need to multiply the 3-bit register offset by `PcdSerialRegisterStride` like `SerialPortReadRegister` and `SerialPortWriteRegister` do.

`CommonPatches.c`:
```
STATIC
UINTN
mPmioRegisterBase = 0;    ///< To be set by PatchSetPciSerialDeviceRegisterBase()

SATIC
UINTN
mPmioRegisterStride = 1;    ///< To be set by PatchSetPciSerialDeviceRegisterBase()
```

`PatchSetPciSerialDeviceRegisterBase`:
```
...
  IN  UINTN  RegisterBase,
  IN  UINTN  RegisterStride
...
mPmioRegisterBase = RegisterBase;
mPmioRegisterStride = RegisterStride;
...
```

`PatchCustomPciSerialPmio`:
```
...
          UINTN Pmio = mPmioRegisterBase + (*WalkerPmio & 7U) * mPmioRegisterStride;
          WalkerPmio[0] = Pmio & 0xFFU;
          WalkerPmio[1] = Pmio >> 8U;
...
```

`PatchCustomPciSerialDevice`:
```
...
  if (mPmioRegisterBase != 0 && (mPmioRegisterBase != 0x3F8U || mPmioRegisterStride > 1) {
...
```

Reply:
> What did I miss?

No, rather I meant that the current doc written by me has almost every detail missing. I will handle this.

> What are the differences between how the patch works and what I described?

For instance, we should not check the value of Misc->Serial->Init to determine whether or not to enable the patch. Instead, we just apply it anyway.

Then regarding the patch, I will commit it later. Thanks for pointing it out.
As for patching `PE_parse_boot_arg`, that sounds crazy to be honest. I think we can just do it in a way that [OcAfterBootCompatLib](https://github.com/acidanthera/OpenCorePkg/tree/master/Library/OcAfterBootCompatLib) does, though I am not sure whether this is feasible. CC @vit9696

Also, in your original issue post, you talked about the boot-arg `msgbuf` - I think now the correct method is to use [DebugEnhancer](https://github.com/acidanthera/DebugEnhancer) instead.

Reply:
> For instance, we should not check the value of Misc->Serial->Init to determine whether or not to enable the patch. Instead, we just apply it anyway.

True. What I meant was: if Serial->Init is not set to true, then the Pcd values don't get changed. If the Pcd values don't get changed, then they will have the default values. If they have the default values, then the patch won't get applied (because the default base address is 0x3f8 and the default stride is 1 and the default device path is 0xff and UseMMIO is default to false.

The PMIO patch should not be applied if RegisterAccessWidth is not 8 since the kernel does only 8 bit accesses to PMIO.

> Also, in your original issue post, you talked about the boot-arg `msgbuf` - I think now the correct method is to use [DebugEnhancer](https://github.com/acidanthera/DebugEnhancer) instead.

Thanks. I'll check it out... `msgbuf` is mentioned in Configuration.pdf. It is a boot-arg that is used in the kernel by bcd_init.c since 10.5.8 (but not 10.4.11 or earlier). The value is used by log_setsize. DebugEnhancer appears to use a value of 10MB for log_setsize. DebugEnhancer has two different patches for setting the size: one for when log_setsize exists, and one for when log_setsize does not exist. I don't know what versions of macOS the latter applies to (it should be commented in the code...)

Does DebugEnhancer require a boot-arg to enable it? It has a boot-arg to disable it `-dbgenhoff`. If DebugEnhancer is enabled, then using `msgbuf=` in boot-args has no effect. Maybe this should be mentioned in Configuration.pdf.

Reply:
Regarding `RegisterStride` - is there any function call to retrieve its value instead of reading it from Misc->Serial part?

Also, I think we should keep `mPmioRegisterBase` and `mPmioRegisterStride` UINT16 since this is what we are able to patch at most.

And we should also remove the `msgbuf` boot-args in docs, and recommend the use of DebugEnhancer instead since it is easier and as of 12.0 it is no longer possible to set this value without DebugEnhancer if I remember correctly. See https://github.com/acidanthera/bugtracker/issues/1916 CC @vit9696

Reply:
> Regarding `RegisterStride` - is there any function call to retrieve its value instead of reading it from Misc->Serial part? Or maybe just call PcdGet functions?

I vote PcdGet. There's a function for RegisterBase because it requires device probing, enabling, and calculation. RegisterStride is much simpler.

> And we should also remove the `msgbuf` boot-args in docs, and recommend the use of DebugEnhancer instead since it is easier and as of 12.0 it is no longer possible to set this value without DebugEnhancer if I remember correctly. See [acidanthera/bugtracker#1916](https://github.com/acidanthera/bugtracker/issues/1916) CC @vit9696

Oh right. My xnu sources don't include macOS 12 yet so I didn't search it. Apple recently updated their open source repositories to include stuff from macOS 12 so I'll update that.

Reply:
Thanks. Regarding `RegisterStride` which is `UINT32`, how can we guarantee that the final patch (UINT16, 2 bytes) will not overflow? I don't think `>> 8U` is enough.

Reply:
> Thanks. Regarding `RegisterStride` which is `UINT32`, how can we guarantee that the final patch (UINT16, 2 bytes) will not overflow? I don't think `>> 8U` is enough.

don't apply the PMIO patch if (UINTN)RegisterBase + 7 * RegisterStride > 0xffff

Reply:
> OK. I updated your new patch. However, this time different values are replaced. Is it expected?

Different than what?

I think the log looks correct.

```
OCAK: Registering PCI serial device PMIO port 0x2008
OCAK: Registering PCI serial device register stride  0x4
OCAK: Matched PMIO serial register base <66 BA FF 03>
OCAK: Matched PMIO serial register base context out <EE>, patching register base
OCAK: Before register base patch <FF 03>
OCAK: After register base patch <24 20>
```

0x2008 + (0x3ff - 0x3f8)*4 = 0x2008 + 28 = 0x2008 + 0x1c = 0x2024.

Reply:
> Different than what?

Without stride (i.e. multiplication arithmetics) it should have always been replaced to 0x2008. But now that you said it is correct, it's all good.

Reply:
Default stride is 1 so the new value for 0x3ff would be 0x2008 + 7 = 0x200f.
Stride cannot be 0.

Reply:
I see, then previously we were applying the patching with stride missing as if it was 0. Thanks for the fix.

Reply:
I think previously were were applying the patch with stride assumed to be 1.
0x2008 | (0x3ff & 7) = 0x200f.